contracts/eosio.system/src/producer_pay.cpp

```diff
@@ -138,34 +164,124 @@ namespace eosiosystem {
       // This is okay because in this case the producer will not get paid anything either way.
       // In fact it is desired behavior because the producers votes need to be counted in the global total_producer_votepay_share for the first time.
 
-      int64_t producer_per_block_pay = 0;
-      if( _gstate.total_unpaid_blocks > 0 ) {
-         producer_per_block_pay = (static_cast<double>(_gstate.perblock_bucket) * prod.unpaid_blocks) / _gstate.total_unpaid_blocks;
-         check( producer_per_block_pay >= 0, "producer per block pay must be greater or equal to 0" );
-      }
+      if (_grewards.activated) {
+         constexpr reward_type reward_types[] = { reward_type::producer, reward_type::standby };
+
+         struct data {
+            int64_t     per_block_pay = 0;
+            double      percent = 0.0;
+            eosio::name account;
+            std::string check_msg;
+            std::string tx_msg;
+         };
+
+         std::map<reward_type, data> info {
+            {  reward_type::producer,
+               { 0,
+                 producer_perc_reward,
+                 bpay_account,
+                 "producer per block pay must be greater or equal to 0",
+                 "producer block pay"
+               }
+            },
+            {  reward_type::standby,
+               { 0,
+                 standby_perc_reward,
+                 spay_account,
+                 "standby per block pay must be greater or equal to 0",
+                 "standby block pay"
+               }
+            }
+         };
+
+         const auto& reward = _rewards.get(owner.value);
+
+         for (auto type: reward_types) {
+            auto& curr_info = info[type];
+
+            if (auto total_unpaid_blocks = _grewards.get_counters(type).total_unpaid_blocks; total_unpaid_blocks > 0) {
+               auto counters = reward.get_counters(type);
+
+               if (counters.selection > 0 && counters.unpaid_blocks > 0) {
+                  const auto perblock_buckets = _gstate.perblock_bucket * curr_info.percent;
```

Comment:
Should this be something like
```
  _const auto perblock_buckets = _counters(reward_type::producer).perblock_bucket;
```
Like I think we need the specific per_block_bucket counter for the producer type yeah? And also, we don't need to multiply by the percentage since that is handled in the fill_buckets method when we split up inflastion.

Reply:
Ohh yes, good catch!, I added the counter to the reward singleton, but I forgot to use it :-O